internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixCustomizedViewOfflinePushRepository.java

```diff
@@ -107,44 +113,31 @@ protected void onCustomizedViewDataChange(RoutingTableSnapshot routingTableSnaps
       // Get number of partitions from Ideal state category in ZK.
-      Map<String, Integer> resourceToPartitionCountMapSnapshot = resourceToIdealPartitionCountMap;
       ResourceAssignment newResourceAssignment = new ResourceAssignment();
       Set<String> resourcesInCustomizedView =
           customizedViewCollection.stream().map(CustomizedView::getResourceName).collect(Collectors.toSet());
 
-      if (!resourceToPartitionCountMapSnapshot.keySet().containsAll(resourcesInCustomizedView)) {
-        LOGGER.info(
-            "Found the inconsistent data between customized view and ideal state of cluster: {}."
-                + " Reading the latest ideal state from zk.",
-            manager.getClusterName());
-        List<PropertyKey> keys = customizedViewCollection.stream()
-            .map(cv -> keyBuilder.idealStates(cv.getResourceName()))
-            .collect(Collectors.toList());
-        try {
-          List<IdealState> idealStates = manager.getHelixDataAccessor().getProperty(keys);
-          refreshResourceToIdealPartitionCountMap(idealStates);
-          resourceToPartitionCountMapSnapshot = resourceToIdealPartitionCountMap;
-          LOGGER.info("Ideal state of cluster: " + manager.getClusterName() + " is updated from zk");
-        } catch (HelixMetaDataAccessException e) {
-          LOGGER.error(
-              "Failed to update the ideal state of cluster: {}. Because we could not access to zk.",
-              manager.getClusterName(),
-              e);
-          return;
-        }
-      }
-
       for (CustomizedView customizedView: customizedViewCollection) {
         String resourceName = customizedView.getResourceName();
-        if (!resourceToPartitionCountMapSnapshot.containsKey(resourceName)) {
-          LOGGER.warn(
-              "Could not find resource: {} in ideal state. Ideal state is up to date, so the resource has been"
-                  + " deleted from ideal state or could not read from zk. Ignore its customized view update.",
-              resourceName);
-          continue;
+        int partitionCount = resourceToPartitionCountMap.getOrDefault(resourceName, -1);
+        if (partitionCount == -1) {
+          String storeName = Version.parseStoreFromKafkaTopicName(resourceName);
+          int version = Version.parseVersionFromVersionTopicName(resourceName);
+          Store store = storeRepository.getStore(storeName);
+          if (store == null) {
+            LOGGER.warn("Cannot find store for resource: {}.", resourceName);
+            continue;
+          }
+          boolean isValidVersion = store.getVersions().stream().anyMatch(v -> v.getNumber() == version);
+          if (!isValidVersion) {
+            LOGGER.warn("Version not found in store for resource: {}.", resourceName);
+            continue;
+          }
+          partitionCount = store.getVersion(version).get().getPartitionCount();
+          resourceToPartitionCountMap.put(resourceName, partitionCount);
```

Comment:
I see puts on this final map but never any remove or clear. Is it a mem leak? Should we instead make the map non-final, build a new one in this function here, and swap the whole map?